rollout/trafficrouting/istio/istio_test.go

```diff
@@ -171,6 +193,41 @@ func TestReconcileVirtualServiceNotFound(t *testing.T) {
 	assert.True(t, k8serrors.IsNotFound(err))
 }
 
+// TestReconcileAmbiguousRoutes tests when we omit route names an there are multiple routes in the VirtualService
+func TestReconcileAmbiguousRoutes(t *testing.T) {
+	obj := unstructuredutil.StrToUnstructuredUnsafe(regularVsvc)
+	client := testutil.NewFakeDynamicClient(obj)
+	ro := rollout("stable", "canary", "vsvc", nil)
+	vsvcLister, druleLister := getIstioListers(client)
+	r := NewReconciler(ro, client, record.NewFakeEventRecorder(), vsvcLister, druleLister)
+	client.ClearActions()
+	err := r.SetWeight(0)
+	assert.Equal(t, "VirtualService spec.http[] must have exactly one route when omitting spec.strategy.canary.trafficRouting.istio.virtualService.routes", err.Error())
+}
+
+// TestReconcileInferredSingleRoute we can support case where we infer the only route in the VirtualService
+func TestReconcileInferredSingleRoute(t *testing.T) {
+	obj := unstructuredutil.StrToUnstructuredUnsafe(singleRouteVsvc)
+	client := testutil.NewFakeDynamicClient(obj)
+	ro := rollout("stable", "canary", "vsvc", nil)
+	vsvcLister, druleLister := getIstioListers(client)
+	r := NewReconciler(ro, client, record.NewFakeEventRecorder(), vsvcLister, druleLister)
+	client.ClearActions()
+	err := r.SetWeight(10)
+	assert.NoError(t, err)
+	actions := client.Actions()
+	assert.Len(t, actions, 1)
+	assert.Equal(t, "update", actions[0].GetVerb())
+
+	// Verify we actually made the correct change
+	vsvcUn, err := client.Resource(istioutil.GetIstioVirtualServiceGVR()).Namespace(ro.Namespace).Get(context.TODO(), "vsvc", metav1.GetOptions{})
+	assert.NoError(t, err)
+	routes, _, _ := unstructured.NestedSlice(vsvcUn.Object, "spec", "http")
+	route := routes[0].(map[string]interface{})
```

Comment:
Shall we add an `assert.Equal` to verify the route name is correctly inferred?

Reply:
Actually the test VirtualService intentionally does not have a route name so there's no name to verify:

```yaml
kind: VirtualService
metadata:
  name: vsvc
  namespace: default
spec:
  gateways:
  - istio-rollout-gateway
  hosts:
  - istio-rollout.dev.argoproj.io
  http:
  - route:
    - destination:
        host: 'stable'
      weight: 100
    - destination:
        host: canary
      weight: 0
```